chips/nrf52840/src/ieee802154_radio.rs

```diff
@@ -767,128 +835,254 @@ impl<'a> Radio<'a> {
+        if self.registers.crcstatus.is_set(Event::READY) {
+            Ok(())
+        } else {
+            Err(ErrorCode::FAIL)
+        }
+    }
+
+    // TODO: RECEIVING ACK FOR A SENT TX IS NOT IMPLEMENTED //
+    // As a general note for the interrupt handler, event registers must still be cleared even when
+    // hardware shortcuts are enabled.
     #[inline(never)]
     pub fn handle_interrupt(&self) {
         self.disable_all_interrupts();
 
-        if self.registers.event_ready.is_set(Event::READY) {
-            self.registers.event_ready.write(Event::READY::CLEAR);
-            self.registers.event_end.write(Event::READY::CLEAR);
-            if self.transmitting.get()
-                && self.registers.state.get() == nrf5x::constants::RADIO_STATE_RXIDLE
-            {
-                self.registers.task_ccastart.write(Task::ENABLE::SET);
-            } else {
-                self.registers.task_start.write(Task::ENABLE::SET);
+        let mut start_task = false;
+        let mut rx_init = false;
+
+        match self.state.get() {
+            // It should not be possible to receive an interrupt while the tracked radio state is OFF
+            RadioState::OFF => panic!("Received interrupt while off"),
+            RadioState::RX => {
+                ////////////////////////////////////////////////////////////////////////////
+                // NOTE: This state machine assumes that the READY_START shortcut is enabled
+                // at this point in time. If the READY_START shortcut is not enabled, the
+                // state machine/driver will likely exhibit undefined behavior.
+                ////////////////////////////////////////////////////////////////////////////
+
+                // Since READY_START shortcut enabled, always clear READY event
+                self.registers.event_ready.write(Event::READY::CLEAR);
+
+                // Completed receiving a packet, now determine if we need to send ACK
+                if self.registers.event_end.is_set(Event::READY) {
+                    self.registers.event_end.write(Event::READY::CLEAR);
+                    let result = self.crc_check();
+
+                    // Unwrap fail = Radio RX Buffer is missing (may be due to receive client not replacing in receive(...) method,
+                    // or some instance in  driver taking buffer without properly replacing).
+                    let rbuf = self.rx_buf.take().unwrap();
+
+                    let frame_len = rbuf[MIMIC_PSDU_OFFSET as usize] as usize - radio::MFR_SIZE;
+                    // Length is: S0 (0 Byte) + Length (1 Byte) + S1 (0 Bytes) + Payload
+                    // And because the length field is directly read from the packet
+                    // We need to add 2 to length to get the total length
+
+                    // 6th bit in 2nd byte of received 15.4 packet determines if sender requested ACK
+                    if rbuf[2] & ACK_FLAG != 0 {
```

Comment:
Shouldn't we only be sending an ACK if the packet requests one AND the CRC is valid?